gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java

```diff
@@ -303,16 +376,21 @@ private boolean shouldMoveToLatestOffset(KafkaPartition partition, SourceState s
     return this.moveToLatestTopics.contains(partition.getTopicName()) || moveToLatestTopics.contains(ALL_TOPICS);
   }
 
-  private WorkUnit createEmptyWorkUnit(KafkaPartition partition, long previousOffset) {
+  private WorkUnit createEmptyWorkUnit(KafkaPartition partition, long previousOffset,
+      Optional<State> topicSpecificState) {
     Offsets offsets = new Offsets();
     offsets.setEarliestOffset(previousOffset);
     offsets.setLatestOffset(previousOffset);
     offsets.startAtEarliestOffset();
-    return getWorkUnitForTopicPartition(partition, offsets);
+    return getWorkUnitForTopicPartition(partition, offsets, topicSpecificState);
   }
 
-  private WorkUnit getWorkUnitForTopicPartition(KafkaPartition partition, Offsets offsets) {
+  private WorkUnit getWorkUnitForTopicPartition(KafkaPartition partition, Offsets offsets,
+      Optional<State> topicSpecificState) {
     Extract extract = this.createExtract(DEFAULT_TABLE_TYPE, DEFAULT_NAMESPACE_NAME, partition.getTopicName());
+    if (topicSpecificState.isPresent()) {
+      extract.addAll(topicSpecificState.get());
```

Comment:
No, there is no functional benefit versus adding it on an `Extract` level vs. a `WorkUnit` level. Although I think it logically makes more sense to add it on an `Extract` level since each topic has a single `Extract`.